Docs: Adding PR template and updating PR guide (#25103)
hashicorp/terraform-provider-azurerm: .github/PULL_REQUEST_TEMPLATE/pull_request_template.md

<!--  All Submissions -->

## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## PR Checklist

> - [ ] Have you followed the guidelines in our [Contributing Documentation](../contributing/README.md)?

Reviewer: should this be I have instead of Have you?
```suggestion
- [ ] I have followed the guidelines in our [Contributing Documentation](../contributing/README.md)
```

Author: I think they all need to be either first person or second person, not some of each. I will change them all to first person and on re-review LMK if it sounds better.

- [ ] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

> - [ ] Have you used a meaningful PR description to help maintainers and other users understand this change and help prevent duplicate work?

Reviewer: this is referring to the title?
```suggestion
- [ ] Have you used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work?
```

Author: I'm not sure. It could really be in reference to either. @jackofallops Do you know if this is a typo or is supposed to be description?

Reviewer: I think it makes more sense to be Title then description given the example is a title

## Changes to existing Resource / Data Source

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them? (This may be covered by linking to an issue above, but may benefit from additional explanation)
- [ ] Have you written new tests for your resource or datasource changes?
- [ ] Have you successfully run tests with your changes locally? If not, please provide details on testing challenges that prevented you running the tests.

## Documentation Changes

- [ ] Documentation is written in International English.
- [ ] Documentation is written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.

## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. It also helps to paste the output from running the acceptance tests. -->

## Related Issue(s)

> Fixes #0000
> ## Change Log

Reviewer: ```suggestion
Fixes #0000

## Change Log
```

> [Changelog Format](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/maintainer-changelog.md)

Reviewer: this should become a sentence? like
```suggestion
Below please provider what should go into the changelog (if anything) conforming to the [Changelog Format documented here] (https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/maintainer-changelog.md)
```

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_resource` - support for the `thing1` property [GH-00000]

<!-- What type of PR is this? -->

> 

Reviewer: maybe add a: This is a: (pleas select all that apply)

- [ ] Bug Fix
- [ ] New Feature
- [ ] Enhancement

> 
> > [!NOTE] If this PR changes meaningfully during the course of review please update the title and description as required.

Reviewer: Should we put this other a heading so it stands out?
```suggestion
## During review and before merge:
 
 If this PR changes meaningfully during the course of review please update the title and description as required.
```

Author: @katbyte Somehow this lost the additional formatting I had that makes it stand out, I added it back, please see below for how it stands out. If you want a header instead of what's below, LMK. 
> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

## PR Checklist

- [ ] I have followed the guidelines in our [Contributing Documentation](../contributing/README.md).
- [ ] I have checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change.
- [ ] I have used a meaningful PR description to help maintainers and other users understand this change and help prevent duplicate work.
Example:
“`resource_name_here` - description of change e.g. adding property `new_property_name_here`”
- [ ] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.

<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->

## New Feature Submissions

- [ ] My submission includes Test coverage as described in the [Contribution Guide](../contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why below)

> ## Changes to existing Resource / Data Source
> 
> - [ ] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
> - [ ] I have written new tests for my resource or datasource changes.
> - [ ] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
> 

Reviewer 2: State migrations can be opened as standalone PRs to fix bugs, or in combination with other changes that could classify more as enhancements to an existing resource, so I'm not sure if it wants it's own header or could just be tagged on here as another check box.
```suggestion
## Changes to existing Resource / Data Source

- [ ] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [ ] I have written new tests for my resource or datasource changes.
- [ ] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a state migration only). I have manually tested the migration path between relevant versions of the provider.
```

Author: I will add this and I am going to bold **state migration only** instead of moving it to its own header. We can try this for now, and if in the future we're not getting the interaction we are aiming for, we can update the template.

## Documentation Changes

- [ ] Documentation is written in International English.
- [ ] Documentation is written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.

> ## Description
> 
> <!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. It also helps to paste the output from running the acceptance tests. -->
> 
> 
> 

Reviewer 2: I think it could be helpful to have a separate heading underneath, specifically for `## Testing Evidence` in order to emphasize the importance of tests/testing for PRs.

From observation I've found that it's easy to forget adding test coverage for changes made to a resource/data source so this will hopefully serve as a strong reminder. This will also let us call out special testing requirements for specific changes, like state migrations.

```suggestion
## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. -->

## Testing Evidence

<!-- Please include your testing evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->

```

<!--  All Submissions -->

## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## PR Checklist

> - [ ] I have followed the guidelines in our [Contributing Documentation](../contributing/README.md).

Reviewer 3: I'm not quite sure this (and other) relative link will work for a PR, as the PR link is like:

`https://github.com/hashicorp/terraform-provider-azurerm/pull/25103`, which will be `https://github.com/hashicorp/terraform-provider-azurerm/pull/contributing/README.md` after evaluated.

Also it's worth mentioning that there is one special guideline in the `contributing` folder is about how to create a PR: https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-opening-a-pr.md. When users check this checkbox, it means they have followed that guideline also, while that guideline by its name shall contain everything in this template, which means we only need to put things there and only leave one checkbox here, which likely make PR authors life much easier and avoid inconsistency between this template and that guideline?

Author: @magodo Thanks for catching the relative links, I have updated them so they should work now.

As for the formatting suggestions, we are discussing ways to reduce checkbox fatigue per your suggestions. I am not sure what changes we will make yet, but please continue to give us feedback once the PR template is live, since it is a living document and we may need to make future changes depending on user experience.
